### PR TITLE
feat: Storybook schematics

### PR DESCRIPTION
**Pull request by isaacplmann**
_[Please make sure you have read the submission guidelines before posting an PR](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md#submit-pr)_

> _Please make sure that your commit message follows our format._

> Example: `fix(nx): must begin with lowercase`

## Current Behavior (This is the behavior we have today, before the PR is merged)
No way to add storybook to a workspace.

## Expected Behavior (This is the new behavior we can expect after the PR is merged)
`@nrwl/storybook` is a set of schematics to programmatically scaffold storybook stories and cypress tests to a ui lib.

TODO:  Add tests for all the schematics.

### CODE REVIEW

File: packages/storybook/src/schematics/storybook-configure/files/.storybook/config.js
Comment (criskrzysiu) [inline packages/storybook/src/schematics/storybook-configure/files/.storybook/config.js]: Is the path correct? Shouldn't it be '../libs'? There is no path like '../src/lib' in nx workspace.

File: packages/storybook/src/schematics/storybook-configure/files/.storybook/tsconfig.json
Comment (criskrzysiu) [inline packages/storybook/src/schematics/storybook-configure/files/.storybook/tsconfig.json]: Again, no such dir as ../src

File: packages/storybook/src/schematics/storybook-configure/storybook-configure.ts
Comment (criskrzysiu) [inline packages/storybook/src/schematics/storybook-configure/storybook-configure.ts]: what happens if we want multiple projects to have storybook? Won't we have a conflict here with the name?

**Comment by isaacplmann** [comment]
Here's a bit more information on what these schematics are intended to generate.

Storybook portion:
1. `.storybook` folder inside of a ui lib
2. `*.stories.ts` file for each `*.component.ts` file declared inside that ui lib
3. A `storybook` builder target in the angular json for that ui lib that runs storybook (just for that lib)

Cypress integration:
1. A `[ui-lib-name]-e2e` app that runs cypress against the storybook for that lib
2. A `*.spec.ts` inside the e2e app for each `*.component.ts` file in the ui lib

File: packages/storybook/src/schematics/storybook-configure/storybook-configure.ts
Comment (isaacplmann) [reply packages/storybook/src/schematics/storybook-configure/storybook-configure.ts]: Sorry, I should have included more information on the goal of this PR.  This schematic is set up to create an isolated storybook instance for a specific ui-lib.  For large monorepos, having a single storybook instance would be un-manageable.

This line doesn't cause a conflict - it creates a `storybook` target for the individual lib (just like each lib has its own `lint` target).  The `run-command` is set to point at the specific `.storybook` folder inside the lib (which only searches that lib for `*.stories.ts` files).

**Comment by isaacplmann** [comment]
The overall goal is to make it quick and easy to run isolated ui tests on individual components using Cypress.

File: packages/storybook/src/schematics/storybook-configure/storybook-configure.ts
Comment (criskrzysiu) [reply packages/storybook/src/schematics/storybook-configure/storybook-configure.ts]: Ok, that does explain a lot. But using storybook only for tests is a bit of waste I think. It would be great to have option of adding different addons, like np storysource etc from some cli prompt when crating story. 

I just integrated storybook in our company repo but went with one storybook per monorepo approach,  but I see some benefits from usage per lib.  

Have You tried how it works if u try to use affected? Does it run only required storybooks?

Comment (isaacplmann) [reply packages/storybook/src/schematics/storybook-configure/storybook-configure.ts]: nx affected doesn't work well for the storybook target, because it runs the first one and the process never finishes to run the next one.  It would make sense for a storybook-build target though.  I haven't written that, but it would be pretty straight forward to add.

nx affected works great for the cypress e2e tests though.  For each affected ui-lib, the storybook is launched, the cypress tests run, then all the processes end and the next e2e test target runs.

I would like to include more options for addons, that just hasn't made in this first pass.  You'll always be able to go into each `.storybook` folder and make whatever customizations you want.

File: packages/storybook/src/schematics/storybook-configure/storybook-configure.ts
Comment (FrozenPandaz) [inline packages/storybook/src/schematics/storybook-configure/storybook-configure.ts]: Can we make a builder for this so that it can accept options?

File: packages/storybook/collection.json
Comment (FrozenPandaz) [inline packages/storybook/collection.json]:
```suggestion
    "configuration": {
```

It feels sort of redundant to have a `storybook-` prefix in a collection of `storybook` schematics.

File: packages/storybook/collection.json
Comment (FrozenPandaz) [inline packages/storybook/collection.json]: This points to `storybook-configure`?

File: packages/storybook/src/schematics/cypress-configure/index.ts
Comment (FrozenPandaz) [inline packages/storybook/src/schematics/cypress-configure/index.ts]: Hm.... I wonder if it would make sense to make `devServerTarget` and `prodDevServerTarget` options for `@nrwl/cypress:cypress-project`

File: packages/storybook/src/schematics/cypress-configure/index.ts
Comment (FrozenPandaz) [inline packages/storybook/src/schematics/cypress-configure/index.ts]: Normally targets CI?

File: packages/storybook/src/schematics/cypress-configure/index.ts
Comment (FrozenPandaz) [inline packages/storybook/src/schematics/cypress-configure/index.ts]: Can't people just pass in `--watch` when running it in dev?

File: packages/storybook/package.json
Comment (FrozenPandaz) [inline packages/storybook/package.json]: I don't know if this one applies. Monorepos are supported with this but this package can be used in a normal Angular CLI workspace

File: packages/storybook/package.json
Comment (FrozenPandaz) [inline packages/storybook/package.json]: I don't know if this one applies. Sure it runs in Node but such is probably true for everything in NPM lol...

File: packages/storybook/package.json
Comment (FrozenPandaz) [inline packages/storybook/package.json]: I don't think this applies. You wouldn't use Storybook with Nest

File: packages/storybook/package.json
Comment (FrozenPandaz) [inline packages/storybook/package.json]: I don't think this applies. Jest testing is separate from storybook and component testing.

File: packages/storybook/package.json
Comment (FrozenPandaz) [inline packages/storybook/package.json]: This is a vague term.

File: packages/storybook/package.json
Comment (FrozenPandaz) [inline packages/storybook/package.json]:
```suggestion
    "@nrwl/cypress": "*",
```

This will get replaced with the same version of `@nrwl/cypress`

File: packages/storybook/src/schematics/storybook-configure/storybook-configure.ts
Comment (FrozenPandaz) [inline packages/storybook/src/schematics/storybook-configure/storybook-configure.ts]: Let's not depend on `schematics/angular` since this should support React and web components as well.

File: packages/storybook/src/schematics/component-story/index.ts
Comment (FrozenPandaz) [inline packages/storybook/src/schematics/component-story/index.ts]: Let's not depend on `schematics/angular` since this should support React and web components as well.

File: packages/storybook/src/schematics/component-story/index.ts
Comment (FrozenPandaz) [inline packages/storybook/src/schematics/component-story/index.ts]: We may want to move this to `@nrwl/angular` since it is specific to Angular.

Then we can do the following:

```bash
ng g @nrwl/angular:story
ng g @nrwl/react:story
```

File: packages/storybook/src/schematics/ng-add/ng-add.ts
Comment (FrozenPandaz) [inline packages/storybook/src/schematics/ng-add/ng-add.ts]: These needs to be a peer dependency?

File: packages/storybook/src/schematics/ng-add/ng-add.ts
Comment (FrozenPandaz) [inline packages/storybook/src/schematics/ng-add/ng-add.ts]: This seems like some plugin... Should we let users add this for themselves if they want to?

File: packages/storybook/src/schematics/storybook-stories/index.ts
Comment (FrozenPandaz) [inline packages/storybook/src/schematics/storybook-stories/index.ts]: Maybe we can move this to `@nrwl/angular` since this is angular specific.

File: packages/storybook/src/utils/versions.ts
Comment (FrozenPandaz) [inline packages/storybook/src/utils/versions.ts]: This seems unused

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

**Comment by FrozenPandaz** [comment]
So excited for Storybook integration! 🎉

File: packages/storybook/src/schematics/storybook-configure/storybook-configure.ts
Comment (isaacplmann) [reply packages/storybook/src/schematics/storybook-configure/storybook-configure.ts]: Not in this PR

File: packages/storybook/src/schematics/cypress-configure/index.ts
Comment (isaacplmann) [reply packages/storybook/src/schematics/cypress-configure/index.ts]: Not in this PR

File: packages/storybook/src/schematics/ng-add/ng-add.ts
Comment (isaacplmann) [reply packages/storybook/src/schematics/ng-add/ng-add.ts]: I'm using this plugin for the core functionality of creating isolated cypress ui tests.  Eventually, I'd like to give people the choice to include other plugins automatically.

**Comment by vsavkin** [comment]
@isaacplmann @FrozenPandaz hey folks, any update on this?

**Comment by isaacplmann** [comment]
@vsavkin The PR is almost there.  I have a e2e test set up, but it keeps silently failing.  When I try to manually recreate what the e2e test is doing, everything works as expected.

File: packages/storybook/src/utils/versions.ts
Comment (mehrad-rafigh) [inline packages/storybook/src/utils/versions.ts]: Why can't we use the same version as storybookAngularVersion? 
As well, can we update to 5.10.1?

**Comment by ThomasBurleson** [comment]
@isaacplmann - earlier you described the intent of these schematics:

Storybook portion:

.storybook folder inside of a ui lib
*.stories.ts file for each *.component.ts file declared inside that ui lib
A storybook builder target in the angular json for that ui lib that runs storybook (just for that lib)
Cypress integration:

A [ui-lib-name]-e2e app that runs cypress against the storybook for that lib
A *.spec.ts inside the e2e app for each *.component.ts file in the ui lib

----

I feel there is another important scenario: 

The ability to generate a 1...n storybook apps the important and load storybook modules for 1...n ui (or feature UI libraries). 

> This would allow designers to quickly validate UI library components and feature UX.

**Comment by Jordan-Hall** [comment]
I've been using Storybook for NX for a bit now. Just never got around to writing the schematics. Personally, i feel this PR is too early to be considered complete and the requirements should be scoped out better before rushing into adding new features.

Rather than having multiple .storybook folders which you are required to configure each time. Would it not be better to add a root level folder which contains the global storybook and extend each UI library from the base config.  I don't know about you but I also configure storybook with the same plugins


root => .storybook => config.js 
```import { addDecorator } from '@storybook/angular';
import { withKnobs } from '@storybook/addon-knobs';

//import styles
import "`bootstrap";

addDecorator(withKnobs);
```

root => .storybook => plugins.js  ```import '@storybook/addon-knobs/register';```

ui-lib => .storybook => plugin.js ```import '../../../.storybook/plugins';```

ui-lib => .storybook => config.js
```import { configure, addDecorator } from '@storybook/angular';
import '../../../.storybook/config'

function loadStories() {
  const req = require.context('../', true, /\.stories\.ts$/);
  req.keys().forEach(filename => req(filename));
}


configure(loadStories, module);
 ```

**Comment by isaacplmann** [comment]
@ThomasBurleson 
That's a great feature.  I think I'll add that as a separate schematic in a future PR.

@Jordan-Hall 
I really like that idea.  That would make the storybook config more inline with the way cypress, jest, karma, etc are handled.

Thanks to both of you for giving some feedback.  I'll put in some more work on this PR tomorrow.

**Comment by criskrzysiu** [comment]
I have such config in one of my playgrounds. Maybe I will try and make
schematic. That would be a great lesson for me as well.

W dniu czw., 15.08.2019 o 16:14 Isaac Mann <notifications@github.com>
napisał(a):

**Comment by Jordan-Hall** [comment]
@isaacplmann I've just noticed we don't have a build option, this PR only has a serve option. It would be good to add support to build the storybook in case we want to deploy it. At least for @ThomasBurleson suggestion of a global storybook. Maybe only offer support to build each lib where a pacakge.json exists.

File: packages/storybook/src/schematics/configuration/files/.storybook/webpack.config.js
Comment (Jordan-Hall) [inline packages/storybook/src/schematics/configuration/files/.storybook/webpack.config.js]: Think storybook should match every stylesheet nx supports. Also, do we require custom scss support? Under the hood, it uses angular cli which already has it supported

File: packages/storybook/src/utils/versions.ts
Comment (Jordan-Hall) [inline packages/storybook/src/utils/versions.ts]: Why are default addons required? It's not needed to set it up manually so shouldn't be required to do it programmatically using schematics

File: packages/storybook/src/schematics/configuration/files/.storybook/webpack.config.js
Comment (isaacplmann) [reply packages/storybook/src/schematics/configuration/files/.storybook/webpack.config.js]: There's some conflict between the angular cli and storybook's webpack config that throws displays an error instead of your storybook content unless we do this custom webpack config workaround.  See: https://github.com/storybookjs/storybook/issues/3593

I haven't tried stylesheets other than css and scss.

File: packages/storybook/src/utils/versions.ts
Comment (isaacplmann) [reply packages/storybook/src/utils/versions.ts]: I'm using the knobs addon in the auto-generated stories.  That's probably the first addon most people will include.

**Comment by Jordan-Hall** [comment]
Just out of curiosity, and i cant wait to see this brought in. Are we not limiting ourselves with this approach to only angular. We also need to look at support react and I think something else needs to be done to get this to work

**Comment by isaacplmann** [comment]
@Jordan-Hall 

Yes, we definitely want to add React support.  I haven't started thinking about how to do that though.  Any ideas?

**Comment by Jordan-Hall** [comment]
I've got a few ideas. It appears the only major difference is how you load/configure stories. Maybe for now just adding it as a required field for which framework to support.
`import { configure } from '@storybook/react';`

along with two extra npm packages react & react-dom

I've currently got your repository checked out looking at a builder for this. I was thinking we could clean up the .storybook folder as long as we pass in the project path.

**Comment by isaacplmann** [comment]
Nice.  I do want to have a builder in place.  Right now I'm working on extending a root configuration like you suggested.

**Comment by Jordan-Hall** [comment]
Which is fine because you may need project-specific configuration which would then require it's own .storybook like your working on now

File: packages/storybook/src/utils/versions.ts
Comment (jaysoo) [reply packages/storybook/src/utils/versions.ts]: Realistically, are there any usages that don't use knobs? I don't think it's a bad default.

File: packages/storybook/src/schematics/configuration/root-files/.storybook/tsconfig.json
Comment (Jordan-Hall) [inline packages/storybook/src/schematics/configuration/root-files/.storybook/tsconfig.json]: This won't work. It's better not to have this base tsconfig. Your better off enforcing this into each individual app, otherwise, all stories will be included.

File: packages/storybook/src/schematics/configuration/root-files/.storybook/config.js
Comment (Jordan-Hall) [inline packages/storybook/src/schematics/configuration/root-files/.storybook/config.js]: Is this for the global storybook? or is this for each individual to extend, If the answer is for individuals you should have this functionality in here

File: packages/storybook/src/schematics/configuration/lib-files/.storybook/config.js__tmpl__
Comment (Jordan-Hall) [inline packages/storybook/src/schematics/configuration/lib-files/.storybook/config.js__tmpl__]: `const req = require.context('../src/', true, /\.stories\.ts$/);`
Think this would be better incase someone has an app level storybook

File: packages/storybook/src/schematics/configuration/root-files/.storybook/tsconfig.json
Comment (isaacplmann) [reply packages/storybook/src/schematics/configuration/root-files/.storybook/tsconfig.json]: This is for running a global storybook that will pick up all stories in the monorepo

File: packages/storybook/src/schematics/configuration/root-files/.storybook/config.js
Comment (isaacplmann) [reply packages/storybook/src/schematics/configuration/root-files/.storybook/config.js]: This is for running a global storybook that will pick up all stories in the monorepo

File: packages/storybook/src/schematics/configuration/lib-files/.storybook/config.js__tmpl__
Comment (isaacplmann) [reply packages/storybook/src/schematics/configuration/lib-files/.storybook/config.js__tmpl__]: Sure, that's a good idea.

File: packages/storybook/src/schematics/configuration/root-files/.storybook/config.js
Comment (Jordan-Hall) [reply packages/storybook/src/schematics/configuration/root-files/.storybook/config.js]: What about extending the base config for the individual one? Wouldn't a better option to add a new application in angular.json for the global storybook to run?

Comment (isaacplmann) [reply packages/storybook/src/schematics/configuration/root-files/.storybook/config.js]: Adding a separate app would work, but the benefit of this `.storybook` folder configuration is that people who are used to running `start-storybook` from the command line will get the global storybook

**Comment by Jordan-Hall** [comment]
@isaacplmann I think I can get a builder sorted for tomorrow night which should work depending on the framework you've selected. Would it be better to wait until after this is merged or should storybook come in as one?

**Comment by Jordan-Hall** [comment]
@isaacplmann How much longer until this will be ready to be merged in? I've got the basic builder working. Just adding all the other customization to allow more flexibility and then update the schematics and write the unit tests. Shouldn't take too much longer, but should add this is the first time working with the angular devkit.

angular.json
```
"storybook": {
          "builder": "@nrwl/storybook:storybook",
          "options": {
            "uiFramework": "@storybook/angular",
            "config": {
              "configFolder": "apps/rina-eas/.storybook"
            }
          }
        }
```

**Comment by isaacplmann** [comment]
My PR won't be merged before Monday certainly. And if you have a mostly
working builder, it would be better to include that with this PR.  Do you
want to open a PR against my branch?

**Comment by Jordan-Hall** [comment]
Hey,

So I builder has run into an issue when using storybook/angular. https://github.com/storybookjs/storybook/pull/7207 We need to wait until version 5.2 come out. I'll work with 5.2 for now so it's ready on release

**Comment by isaacplmann** [comment]
@Jordan-Hall , can you share your in progress code?  I have more time this week and can help work around the issue you ran into.

**Comment by Jordan-Hall** [comment]
Will do, I've added the whole patch to storybook. It build the code fine just not built it as it ends the feed straight away. Need to start in a child process and on end then close everything. On way to hotel, and revisiting it and I'll push in the morning

**Comment by Jordan-Hall** [comment]
@isaacplmann  sorry for the delay, PR https://github.com/isaacplmann/nx/pull/1/files. Working on changing the schematics over to new builder and writing unit tests.

Please note the hotfix folder is design to fix the issue with the angular package. Storybook doesn't currently work with angular 8

**Comment by Jordan-Hall** [comment]
@isaacplmann Any update on your end? I'm able to help again this weekend to get this PR passing

**Comment by isaacplmann** [comment]
@Jordan-Hall, I got the builder working in the e2e tests without the hotfix code.  Thanks for your help!

**Comment by Jordan-Hall** [comment]
That's great news, i had to use hotfix because they where waiting on PR and upgrade

File: packages/storybook/package.json
Comment (kroeder) [reply packages/storybook/package.json]: I'm not familiar with what should be added as `keywords` but this may help: 

in a default angular project using `sb init` to bootstrap Storybook the `.storybook/tsconfig.json` has `types: ["node"]` https://github.com/storybookjs/storybook/blob/69d9c98d3be0c845c0b8bab96a099c4392364a78/lib/cli/generators/ANGULAR/template/.storybook/tsconfig.json

I added it myself but don't right away know why I added it - if this is of interest I can look it up for you, just let me know. So yes, kinda everything is "node" in npm but Storybook explicitly needs it - may be a reason to add it here?

File: packages/storybook/package.json
Comment (kroeder) [inline packages/storybook/package.json]: fyi https://github.com/DefinitelyTyped/DefinitelyTyped/pull/38447#issuecomment-541112865 a list of deleted types from @types that are included in storybook 5.2.x

**Comment by kroeder** [comment]
Impressive PR! I just scrolled through the changes and left a couple of comments

If you need any help or having trouble integrating Storybook into nx as a schematic let me know or open up an issue in the Storybook repo! I'm really looking forward to have this in nx!

File: packages/storybook/src/schematics/configuration/root-files/.storybook/webpack.config.js
Comment (kroeder) [inline packages/storybook/src/schematics/configuration/root-files/.storybook/webpack.config.js]: I'm curious, why is an initial `webpack.config.js` needed for loading css?
As a reference, the webpack config of `@storybook/angular`: https://github.com/storybookjs/storybook/blob/next/app/angular/src/server/framework-preset-angular.ts

File: packages/storybook/src/schematics/configuration/lib-files/.storybook/tsconfig.json
Comment (kroeder) [inline packages/storybook/src/schematics/configuration/lib-files/.storybook/tsconfig.json]: As said in another comment, I had to add `types: ["node"]` but I'm not sure anymore why I needed it. Would it help to figure out the reason? 

`tsconfig.json` from the Storybook CLI: https://github.com/storybookjs/storybook/blob/next/lib/cli/generators/ANGULAR/template/.storybook/tsconfig.json

File: packages/storybook/src/schematics/configuration/root-files/.storybook/webpack.config.js
Comment (Jordan-Hall) [reply packages/storybook/src/schematics/configuration/root-files/.storybook/webpack.config.js]: Because that would be limiting us to storybook/angular

Comment (kroeder) [reply packages/storybook/src/schematics/configuration/root-files/.storybook/webpack.config.js]: That's true, haven't thought about this 🙂  

In case you didn't knew there is also a presets API https://storybook.js.org/docs/presets/writing-presets/

It's basically about putting the code out of the schematic/bootstrapped storybook inside an internal preset, lets name it `@nrwl/storybook/preset-angular`, `@nrwl/storybook/preset-react` , etc , and maintain the configuration from the inside of your package instead of updating it via e.g. `ng update` on the end-users side 

Those presets can also be overwritten by the end-user, if needed.
The cons: The user doesn't know what happens behind a preset (which is kind of the angular way of things but not entirely true for e.g. react I guess?)

**Comment by kristofdegrave** [comment]
Any idea when this PR will merged and released? I'm looking forward on using this functionality in my projects.

**Comment by isaacplmann** [comment]
@kristofdegrave, @vsavkin is planning to review it soon, but he just got back from a conference so it might take him a bit.

File: packages/cypress/src/schematics/cypress-project/files/cypress.json
Comment (vsavkin) [inline packages/cypress/src/schematics/cypress-project/files/cypress.json]: Why is this necessary?

**Review by vsavkin — CHANGES_REQUESTED** [review]

**Comment by Jordan-Hall** [comment]
Did you add schematics for react as the builder supports it

**Comment by isaacplmann** [comment]
Working on that next

**Pull request merged**
